### PR TITLE
expression: add EvalXXX functions for Constant and Column

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit refines struct Constant to support new typeinferer.
@coocood PTAL

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -176,6 +250,161 @@ func (c *Constant) HashCode() []byte {
...
+	*Constant
+}
+
+// EvalString returns string representation of StringConstant.
+func (c *StringConstant) EvalString(row []types.Datum, sc *variable.StatementContext) (string, bool, error) {
+	val := c.Value
+	if val.IsNull() {
+		return "", true, nil
+	}
+	res, err := val.ToString()
```
Comment (coocood) [inline expression/expression.go]: `val.GetString()` will do.

File: expression/expression.go [hunk 2]
```diff
@@ -91,6 +93,38 @@ func EvalBool(expr Expression, row []types.Datum, ctx context.Context) (bool, er
 	return i != 0, nil
 }
 
+func evalExpr(expr Expression, tp types.TypeClass, row []types.Datum, sc *variable.StatementContext) (interface{}, bool, error) {
```
Comment (coocood) [inline expression/expression.go]: Avoid using `interface{}`.
We can just return `(types.Datum, error)`.
Caller can call `d.GetInt64()`, `d.GetString()` ...
Use name `evalExprToTypeClass` is better.

Comment (coocood) [inline expression/expression.go]: Add comments about those functions that are temporarily used, and it should be removed in the future.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @shenli

**Comment by XuHuaiyu** [comment]
PTAL @shenli

**Comment by XuHuaiyu** [comment]
It would involve too many code changes if we merge this PR,
we'd better refactor Constant and Column later.

**Comment by XuHuaiyu** [comment]
@shenli @coocood PTAL

File: expression/expression.go [hunk 3]
```diff
@@ -91,6 +92,45 @@ func EvalBool(expr Expression, row []types.Datum, ctx context.Context) (bool, er
 	return i != 0, nil
 }
 
+// evalExprToTypeClass evaluates `expr` to specified `tp`.
+// This function is temporarily used to guarantee Constant and Column implement interface Expression.
+func evalExprToTypeClass(expr Expression, tp types.TypeClass, row []types.Datum, sc *variable.StatementContext) (d types.Datum, err error) {
```
Comment (coocood) [inline expression/expression.go]: define 4 evel function for each type class.

**Comment by XuHuaiyu** [comment]
@coocood  PTAL

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@hanfei1991 PTAL

**Comment by XuHuaiyu** [comment]
@shenli @hanfei1991 PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**